fix issue #8268 (joinPaths); add lots of tests  and make use of use runnableExamples

timotheecour opened the pull request:
/cc @dom96 

* fixes https://github.com/nim-lang/Nim/issues/8268 by adding an optional `absOverrides` to `joinPath` (by default, same behavior as before so no breaking change)
* correctly handles boundary bw head and tail so that there's a unique dirsep between
* adds `normalizePathEnd` (also useful in other places, can be refactored later)
* adds lots of tests (and fixes other things that were broken with joinPaths)
* add `countWhile` and `rootPrefixLength` (simplifies and fixes a lot of code)

## eg motivation
`joinPath(foo, bar)`  is often a bug when `bar` is absolute, for eg:
https://github.com/genotrance/nimterop/commit/a8bb2dc01f99a2866f586f6e1491a512ce9b3cfa#r31391178
> don't joinPath if path.isAbsolute

Varriount commented:
Why is normalizePathEnd public? Even if required by other procedures in the os module, it seems somewhat niche.

timotheecour commented on lib/pure/ospaths.nim:
```diff
@@ -157,64 +157,124 @@ const
+      doAssert joinPath("usr///", "//lib") == "usr/lib" ## `//` gets compressed
+      doAssert joinPath("//", "lib") == "/lib" ## ditto
+    when defined(Windows):
+      doAssert joinPath("""C:\foo""", """D:\bar""") == """C:\foo\bar"""
```
done, thx for the tip!

timotheecour commented on lib/pure/ospaths.nim:
```diff
@@ -157,64 +157,124 @@ const
-  ##   assert joinPath("usr/", "/lib") == "usr/lib"
+  ## If ``tail`` is absolute or ``head`` is empty, returns ``tail``. If
+  ## ``tail`` is empty returns ``head``. Else, returns the concatenation with
+  ## normalized spearator between ``head`` and ``tail``.
```
done

timotheecour commented:
> Why is normalizePathEnd public? Even if required by other procedures in the os module, it seems somewhat niche.

I think it's useful, I would want to use it in my code

timotheecour commented:
PTAL, now green

dom96 commented on lib/pure/ospaths.nim:
```diff
@@ -157,64 +158,173 @@ const
     ## for example, the '.' in ``os.nim``.
 
+# TODO: MOVE to strutils, sequtils, or algorithms
+proc countWhile*(str: string, pred : proc(a:char):bool, prefix = true): int =
```
You can use ``parseutils.parseWhile`` for this, or just ``str.len-str.replace("/", "").len``

If you want more complicated counting then just use a for loop...

timotheecour replied:
`parseWhile` is not as flexible (eg I need arbitrary lambda; and I need to go from both ends).

`str.len-str.replace("/", "").len` won't work with my other usages of `countWhile`.

> If you want more complicated counting then just use a for loop...

`countWhile` leads to simpler code

dom96 commented on lib/pure/ospaths.nim:
```diff
@@ -157,64 +158,173 @@ const
+  return lenS
+
+# TODO: once slices are supported, define instead `rootPrefix`
+proc rootPrefixLength*(path: string) : int =
```
Once slices are supported? Just return the prefix and call ``len`` on it.

timotheecour replied:
`rootPrefixLength` avoids allocating. If/when we have slices, this wouldn't be a problem; until then it is.

dom96 commented on lib/pure/ospaths.nim:
```diff
@@ -157,64 +158,173 @@ const
+      doAssert r"C:".rootPrefixLength == 2
+      doAssert r"C:\foo".rootPrefixLength == 3
+      doAssert r"//foo".rootPrefixLength == 2
+      doAssert r"C:\\foo".rootPrefixLength == 4
```
As far as I'm concerned this is abusing ``runnableExamples`` as exhaustive testing again. Examples should be one or two (one per platform). Not a long list of edge cases.

timotheecour replied:
fixed; moved out from runnableExamples

timotheecour replied:
fixed; moved out of runnableExamples

dom96 commented on lib/pure/ospaths.nim:
```diff
@@ -157,64 +158,173 @@ const
+    if path[0] == ':':
+      result = 0
+    else:
+      result = countWhile(path, a => a != ':')
```
Is ``macos`` pre-mac-os-x only? If it covers latest macOS then this looks wrong to me.

citycide replied:
IIRC `macos` is there legacy classic Mac, `macosx` is, well, Mac OS X. This is kind of confusing though because since 10.12 Sierra they rebranded it as macOS.

dom96 replied:
yeah, this is far too confusing.

timotheecour replied:
what is too confusing? the code in PR or `macos`? I already suggested removing support for `macos` https://forum.nim-lang.org/t/4039 "should we keep supporting macos? (as opposed to macosx)"

dom96 commented on lib/pure/ospaths.nim:
```diff
@@ -157,64 +158,173 @@ const
+      doAssert "/".isAbsolute
+      doAssert(not "a/".isAbsolute)
+    when defined(Windows):
+      doAssert "C:\\foo".isAbsolute
```
Too many examples here too.

timotheecour replied:
fixed

dom96 commented on lib/pure/ospaths.nim:
```diff
@@ -157,64 +158,173 @@ const
-  ##   assert joinPath("", "lib") == "lib"
-  ##   assert joinPath("", "/lib") == "/lib"
-  ##   assert joinPath("usr/", "/lib") == "usr/lib"
+  ## If ``tail`` is absolute and ``absOverrides`` is true, or ``head`` is empty,
```
I'm not really sure about the use case for this.

timotheecour replied:
rationale is given in #8268 : most languages use the behavior absOverrides=true whereas before this PR nim was using the equivalent of absOverrides=false; so with this PR, 0 code is broken but user has option of using absOverrides=true

dom96 commented:
I also think that ``normalizePathEnd`` is niche and shouldn't be there. I'm unsure about what it's purpose is in ``joinPath``.

dom96 commented:
All in all there is too much going on here. I would prefer if you just created a simple PR that fixes the underlying issue.

timotheecour commented:
> I'm pretty sure for most purposes normalizePathEnd can just be if path[^1] == '/': path[0 .. ^2]

doesn't work with `path = "foo//"` etc. The way I wrote it via `normalizePathEnd` is robust and can be reused instead of being wrong some of the time; I found a ton of bugs in ospaths module because of lack of encapsulation/helper functions like these.

> I also think that normalizePathEnd is niche and shouldn't be there. I'm unsure about what it's purpose is in joinPath.

I have other upcoming PR's that fix pre-existing bugs with ospaths that rely on `normalizePathEnd`, `countWhile`, `rootPrefixLength`. Having these helpers makes code simpler, more reusable, less buggy.

I'm willing to address the comments that make sense to me:
* moving some examples to tos.nim instead of `runnableExamples`
* moving `countWhile` to strutils.nim
* any other comment that makes sense to me

but I'm not ok with removing these helpers. Let me know if you want me to go ahead with this plan otherwise I'll just apply these fixes in a private library and someone else can fix these issues in ospaths (there are many, see https://github.com/nim-lang/Nim/issues/8177).

Araq commented:
> but I'm not ok with removing these helpers. Let me know if you want me to go ahead with this plan otherwise I'll just apply these fixes in a private library and someone else can fix these issues in ospaths (there are many, see a list #8177 + others I haven't reported).

I'm not ok that you have to rewrite large parts of mostly working code with your own ideas of how things should be done in Nim/D and leaving a track record of well meaning, but time-consuming RFCs/proposals as a side effect that take away from our manpower. I really appreciate your help, but if in the end your help makes us all more unproductive than we were before, I have to decline it.

Let's assume ``ospaths.nim`` has 20 bugs (believe it or not, many people use it effectively, today!). Every fix is maybe 5 lines of code, every fix can be a tiny PR that keeps the CIs green. That means the core dev get to review 20 little PRs that can be accepted in maybe one hour. These PRs also probably wouldn't add new dependencies between the stdlib modules, btw.

Your way of working: Write 20 RFCs instead, half of them about perceived more general language deficiencies. And here is the bad part of it: More/different language features just **move** problems around most of the time. For example, I don't want to use D's superior const system, I know from experience it's not worth it, given its complexity, it doesn't find enough real bugs. (Same holds for Nim's ``tags`` effect system.) I don't want to use D's superior "copy-free" slicing system either, it introduces problematic aliasing.

timotheecour commented:
/cc @Araq PTAL 
* no new dependency added (removed the `import sugar` I had added)
* no new public API (made countWhile private; and incorporated all comments from https://github.com/nim-lang/Nim/pull/8673 except `s/proc/set/` ; it's private so shd be fine)
* runnableExamples kept to reasonable size
* moved tests to tests/stdlib/tospaths.nim except ones that refer to private symbols
* some code had to be moved in this PR (eg `normalizePathEnd` which was already committed from a previous PR) ; you can verify for yourself that the move doesn't introduce changes using `git diff --color-moved` ; I tried using `{.reorder:on.}` but it exposed bugs in `reorder:on` implementation (will send a bug for that) ; also tried using function declaration but also led to other issues; so I just moved up some code.

PS: I'm not gonna attempt defending my POV and instead will focus here on moving forward with this PR

Araq commented:
Needs to be redone, ``os.nim`` changed significantly.
